src/lib/storage/zero_suppression/base_zero_suppression_vector.hpp

```diff
@@ -0,0 +1,129 @@
+ * - the vector, which is returned by the encoder and contains the encoded data (base class: BaseZeroSuppressionVector)
+ * - the iterator, for sequentially decoding the vector (base class: BaseZeroSuppressionIterator)
+ * - the decoder, which implements point access into the vector (base class: BaseZeroSuppressionDecoder)
+ *
+ * The iterators and decoders are created via virtual and non-virtual methods of the vector interface.
+ */
+class BaseZeroSuppressionVector : private Noncopyable {
+ public:
+  virtual ~BaseZeroSuppressionVector() = default;
+
+  /**
+   * @brief Returns the number of elements in the vector
+   */
+  virtual size_t size() const = 0;
+
+  /**
+   * @brief Returns the physical size of the vector
+   */
+  virtual size_t data_size() const = 0;
+
+  virtual ZsVectorType type() const = 0;
+
+  virtual pmr_vector<uint32_t> decode() const = 0;
+
+  virtual std::unique_ptr<BaseZeroSuppressionDecoder> create_base_decoder() const = 0;
+
+  virtual std::shared_ptr<BaseZeroSuppressionVector> copy_using_allocator(
+      const PolymorphicAllocator<size_t>& alloc) const = 0;
+};
+
+/**
+ * You may use this iterator facade to implement iterators returned
+ * by ZeroSuppressionVector::cbegin() and ZeroSuppressionVector::cend()
+ */
+template <typename Derived>
+using BaseZeroSuppressionIterator = boost::iterator_facade<Derived, uint32_t, boost::forward_traversal_tag, uint32_t>;
+
+/**
+ * @brief Implements the non-virtual interface of all vectors
+ *
+ * Sub-classes must implement all method starting with `_on_`.
+ */
+template <typename Derived>
+class ZeroSuppressionVector : public BaseZeroSuppressionVector {
+ public:
+  /**
+   * @defgroup Non-virtual interface
+   * @{
+   */
+
+  /**
+   * @brief Returns a vector specific decoder
+   * @return a unique_ptr of subclass of BaseZeroSuppressionDecoder
+   */
+  auto create_decoder() const { return _self()._on_create_decoder(); }
+
+  /**
+   * @brief Returns an iterator to the beginning
+   * @return a constant forward iterator returning uint32_t
+   */
+  auto cbegin() const { return _self()._on_cbegin(); }
+
+  /**
+   * @brief Returns an iterator to the end
+   * @return a constant forward iterator returning uint32_t
+   */
+  auto cend() const { return _self()._on_cend(); }
+  /**@}*/
+
+ public:
+  /**
+   * @defgroup Virtual interface implementation
+   * @{
+   */
+
+  size_t size() const final { return _self()._on_size(); }
+  size_t data_size() const final { return _self()._on_data_size(); }
+
+  ZsVectorType type() const final { return get_zs_vector_type<Derived>(); }
+
+  pmr_vector<uint32_t> decode() const final {
+    auto decoded_vector = pmr_vector<uint32_t>{};
+    decoded_vector.reserve(size());
+
+    auto it = cbegin();
+    const auto end = cend();
+    for (; it != end; ++it) {
+      decoded_vector.push_back(*it);
+    }
+
+    return decoded_vector;
+  }
+
+  std::unique_ptr<BaseZeroSuppressionDecoder> create_base_decoder() const final {
```

Comment:
This file does not explain how it is different from `create_decoder`.

Reply:
Yes it does. `create_base_decoder` returns `std::unique_ptr<BaseZeroSuppressionDecoder>` and in the comment for `create_decoder` the following is written: `@return a unique_ptr of subclass of BaseZeroSuppressionDecoder`

Reply:
Maybe my comment was unclear here. If `create_decoder` gives me something that can be implicitly converted to `std::shared_ptr<BaseZeroSuppressionDecoder>`, why would I need this method?

Reply:
Because I don’t want the base class but the sub-class so that the compiler can omit the virtual method call.

Reply:
Do you have any numbers on this? As we should only create one decoder per chunk, this appears to be a micro-optimization that leads to ambiguity between the two methods.

Reply:
Max: I didn't mean the virtual method call of `create_base_decoder` but every call to `get()` on the decoder, which happens for every position in the position list. Using `create_decoder` you get a sub-class of `BaseVectorDecompressor` where `get()` is declared final so that the compiler can emit the virtual method call.